File: 07-trees/00572-Subtree_Of_Another_Tree/00572-subtree_of_another_tree.go
// 572: Subtree Of Another Tree
// https://leetcode.com/problems/subtree-of-another-tree/

package main

import "fmt"

type TreeNode struct {
    val any
    left *TreeNode
    right *TreeNode
}

type Tree struct {
    root *TreeNode
}

func creator(values []any, root **TreeNode, i, n int) *TreeNode {
    if n==0 {return nil}
    if i<n {
        temp := &TreeNode{values[i], nil, nil}
        *root = temp
        (*root).left = creator(values, &((*root).left), 2*i+1, n);
        (*root).right = creator(values, &((*root).right), 2*i+2, n);
    }
    return *root;
}

func createTree(root **TreeNode, inputs []any) {
    creator(inputs, root, 0, len(inputs))
}

func showTree(root *TreeNode) {
    var q []*TreeNode;
    var result [][]interface{}
    var c []interface{}
    if root==nil { fmt.Println("Empty !"); return; }
    q = append(q, root)
    q = append(q, nil)
    for len(q)!=0 {
        t := q[0]
        q = q[1:]
        if t==nil {
            result = append(result, c)
            c = make([]interface{}, 0)
            if len(q) > 0 {q = append(q, nil)}
        } else {
            c = append(c, t.val)
            if t.left!=nil {q = append(q, t.left)}
            if t.right!=nil {q = append(q, t.right)}
        }
    }

    fmt.Print("["); for _, x := range result {
        fmt.Print("["); for _, y := range x {
            if y==nil { fmt.Print("NULL,"); continue; }
            fmt.Print(y,",")
        }; fmt.Print("\b],")
    }; fmt.Println("\b]")
}

func isSame(root, subRoot *TreeNode) bool {
    if root == nil && subRoot == nil {return true}
    if root == nil || subRoot == nil {return false}
    if root.val != subRoot.val {return false}
    return isSame(root.left, subRoot.left) && isSame(root.right, subRoot.right)
}

// SOLUTION
func isSubtree(root, subRoot *TreeNode) bool {
    if root == nil {return false}
    if isSame(root, subRoot) {return true}
    return isSubtree(root.left, subRoot) || isSubtree(root.right, subRoot)
}



func main() {
    tree, subTree := Tree{}, Tree{}

    // INPUT
    rn, srn := []any{3,4,5,1,2}, []any{4,1,2}
    createTree(&tree.root, rn)
    createTree(&subTree.root, srn)

    // OUTPUT
    result := isSubtree(tree.root, subTree.root)
    fmt.Println(result)
}

File: 07-trees/00572-Subtree_Of_Another_Tree/00572-subtree_of_another_tree.java
// 572: Subtree Of Another Tree
// https://leetcode.com/problems/subtree-of-another-tree/


import java.util.Queue;
import java.util.ArrayList;
import java.util.LinkedList;


class Solution {
    public static class TreeNode {
        Integer val;
        TreeNode left;
        TreeNode right;
        TreeNode() {}
        TreeNode(Integer val) { this.val = val; }
        TreeNode(Integer val, TreeNode left, TreeNode right) { this.val = val; this.left = left; this.right = right; }
    }

    public static class Tree {
        TreeNode root;
        Tree() { this.root = null; }
    }

    private static TreeNode creator(Integer[] values, TreeNode root, int i, int n) {
        if (n==0) return null;
        if (i<n) {
            TreeNode temp = new TreeNode(values[i]);
            root = temp;
            root.left = creator(values, root.left, 2*i+1, n);
            root.right = creator(values, root.right, 2*i+2, n);
        }
        return root;
    }

    public static void createTree(Tree tree, Integer[] inputs) {
        tree.root = creator(inputs, tree.root, 0, inputs.length);
    }

    public static void showTree(TreeNode root) {
        Queue<TreeNode> q = new LinkedList<>();
        ArrayList<ArrayList<Integer>> result = new ArrayList<>();
        ArrayList<Integer> c = new ArrayList<>();
        if (root == null) { System.out.println("Empty !"); return; }
        q.add(root);
        q.add(null);
        while (!q.isEmpty()) {
            TreeNode t = q.poll();
            if (t==null) {
                result.add(c);
                c = new ArrayList<>();
                if (q.size() > 0) q.add(null);
            } else {
                c.add(t.val);
                if (t.left!=null) q.add(t.left);
                if (t.right!=null) q.add(t.right);
            }
        }

        System.out.print("["); for (var x : result) {
            System.out.print("["); for (var y : x) {
                if (y==null) { System.out.print("NULL,"); continue; }
                System.out.print(y+",");
            } System.out.print("\b],");
        } System.out.println("\b]");
    }


    public boolean isSame(TreeNode root, TreeNode subRoot) {
        if (root == null && subRoot == null) return true;
        if (root == null || subRoot == null) return false;
        if (root.val != subRoot.val) return false;
        return isSame(root.left, subRoot.left) && isSame(root.right, subRoot.right);
    }

    // SOLUTION
    boolean isSubtree(TreeNode root, TreeNode subRoot) {
        if (root == null) return false;
        if (isSame(root, subRoot)) return true;
        return isSubtree(root.left, subRoot) || isSubtree(root.right, subRoot);
    }

    

    public static void main(String[] args) {
        Solution o = new Solution();
        Tree tree = new Tree();
        Tree subTree = new Tree();

        // INPUT
        Integer[] rn = {3,4,5,1,2};
        createTree(tree, rn);
        Integer[] srn = {4,1,2};
        createTree(subTree, srn);

        // OUTPUT
        var result = o.isSubtree(tree.root, subTree.root);
        System.out.println(result);
    }
}